# Homework 4: Data Transformation with dplyr - Part 2 (Mutate, Summarize, Group_by, Count)

Welcome to Homework 4! This assignment builds on your foundational dplyr skills by introducing advanced data transformation techniques essential for business analytics.

Name: Charles Moten-Lopez

Due Date: September 21st, 2025

## Learning Objectives

By completing this homework, you will:
- **Master mutate()**: Create new variables and calculated fields for business insights
- **Apply summarize()**: Generate aggregate statistics and key performance indicators (KPIs)
- **Utilize group_by()**: Perform grouped analysis across business dimensions
- **Implement count()**: Conduct frequency analysis and cross-tabulations
- **Develop business intelligence**: Combine functions to create comprehensive analytics

## Business Context

You'll be working with real company sales data to perform the type of analysis that drives business decisions. This includes:
- **Financial Analysis**: Calculate profit margins, ROI, and efficiency metrics
- **Performance Segmentation**: Categorize transactions and customers by performance
- **Regional Analysis**: Compare performance across geographic regions
- **Product Analysis**: Evaluate product category performance
- **Time-based Analysis**: Identify trends and seasonal patterns
- **Sales Rep Performance**: Assess individual representative effectiveness

## Dataset Information

You'll be working with `company_sales_data.csv` which contains:
- **Sales transactions** with revenue, cost, and unit data
- **Geographic information** with regional breakdowns
- **Product categories** for portfolio analysis
- **Sales representative** performance data
- **Date information** for trend analysis
- **Customer metrics** for segmentation analysis

## Instructions

Complete each section by writing R code in the designated areas. Focus on creating clean, well-commented code that demonstrates your understanding of both the technical concepts and business applications.

**Remember**: In real business analytics, you're not just manipulating data - you're uncovering insights that drive strategic decisions!




## Part 1: Setup and Data Import

**Task**: Load the necessary packages and import your dataset.

**Business Context**: Every analysis begins with proper data setup. In professional environments, data scientists spend significant time ensuring data is properly loaded and validated before analysis begins.

**What you need to do**:
1. Load the `tidyverse` package (which includes dplyr)
2. Import the `company_sales_data.csv` file 
3. Examine the structure and basic properties of your data
4. Perform initial data validation

**Hint**: Use `read_csv()` for better data type detection, and always examine your data structure with `str()`, `summary()`, and `head()` functions.

In [1]:
# Load required packages
# TODO: Load tidyverse
library(tidyverse)

# Import the company sales data
# TODO: Read the company_sales_data.csv file into a variable called company_data
company_data <- read_csv("company_sales_data.csv")

# Examine the dataset structure
# TODO: Use str() to examine the data structure
str(company_data)

# Display the first few rows
# TODO: Use head() to show the first 6 rows
head(company_data)

# Generate summary statistics
# TODO: Use summary() to get basic statistics for all columns
summary(company_data)

# Check dataset dimensions
# TODO: Display the number of rows and columns using nrow() and ncol()
nrow(company_data)
ncol(company_data)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.1     ✔ stringr   1.5.2
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


ERROR: Error: 'company_sales_data.csv' does not exist in current working directory ('/workspaces/Assignment-4-Data-Transformation-with-dplyr---Part-2').


## Part 2: Creating New Variables with mutate()

**Task**: Use `mutate()` to create new calculated fields for business analysis.

**Business Context**: Raw data rarely tells the complete story. Business analysts must create derived metrics like profit margins, efficiency ratios, and performance categories to generate actionable insights.

**Variables to create**:
1. **Profit**: Revenue minus Cost
2. **Profit_Margin**: (Profit / Revenue) × 100
3. **Cost_Ratio**: (Cost / Revenue) × 100  
4. **Revenue_Per_Unit**: Revenue divided by Units_Sold
5. **Cost_Per_Unit**: Cost divided by Units_Sold
6. **ROI**: (Profit / Cost) × 100

**Hint**: Use the pipe operator (`%>%`) to chain multiple mutate operations together. Remember to create meaningful variable names that clearly indicate what each metric represents.

In [ ]:
# Create basic financial metrics using mutate()
company_data <- company_data %>%
  mutate(
    # TODO: Create Profit (Revenue - Cost)
    Profit = (Revenue - Cost),
    # TODO: Create Profit_Margin ((Profit / Revenue) * 100)
    Profit_Margin = ((Profit / Revenue) * 100),
    # TODO: Create Cost_Ratio ((Cost / Revenue) * 100)
    Cost_Ratio = ((Cost / Revenue) * 100),
    # TODO: Create Revenue_Per_Unit (Revenue / Units_Sold)
    Revenue_Per_Unit = (Revenue / Units_Sold),
    # TODO: Create Cost_Per_Unit (Cost / Units_Sold)
    Cost_Per_Unit = (Cost / Units_Sold),
    # TODO: Create ROI ((Profit / Cost) * 100)
    ROI = (Profit / Cost) * 100
  )

# Display the new financial metrics
# TODO: Select and display columns: Revenue, Cost, Profit, Profit_Margin, ROI
company_data %>%
select(Revenue, Cost, Profit, Profit_Margin, ROI)

Revenue  Cost     Profit   Profit_Margin ROI      
1   20750.92 12253.36  8497.56 40.95028       69.34882
2   32359.98 24595.20  7764.78 23.99501       31.57031
3   39268.40 23291.09 15977.31 40.68745       68.59838
4   28865.09 12428.74 16436.35 56.94197      132.24470
5    3932.36  1778.18  2154.18 54.78084      121.14522
6   48209.75 26052.04 22157.71 45.96106       85.05173
7   33055.35 20432.24 12623.11 38.18780       61.78035
8   49856.54 37118.63 12737.91 25.54913       34.31676
9   13477.13  8685.85  4791.28 35.55119       55.16190
10  21188.96 16757.06  4431.90 20.91608       26.44796
11  24474.26  7469.31 17004.95 69.48096      227.66427
12  48997.22 26134.68 22862.54 46.66089       87.47970
13   8079.98  3889.00  4190.98 51.86869      107.76498
14  15606.39  6063.92  9542.47 61.14463      157.36471
15  30287.40 12633.44 17653.96 58.28813      139.73993
16  37277.72 15418.04 21859.68 58.64007      141.77989
17  38674.17 27501.44 11172.73 28.88939       40.62598
18  44751.74 26480.82 18270.92 40.82728       68.99681
19  38301.63 19173.66 19127.97 49.94036       99.76170
20   2365.46  1323.14  1042.32 44.06416       78.77624
21   8713.33  6309.40  2403.93 27.58911       38.10077
22  35684.60 20041.90 15642.70 43.83600       78.04999
23  48747.95 17027.61 31720.34 65.07010      186.28768
24   5931.50  3693.96  2237.54 37.72300       60.57294
25   5568.06  3489.03  2079.03 37.33850       59.58762
26  47637.19 29136.66 18500.53 38.83632       63.49571
27  41576.83 31637.33  9939.50 23.90634       31.41700
28   5195.86  2110.14  3085.72 59.38805      146.23295
29  32028.51 25582.75  6445.76 20.12507       25.19573
30   9565.87  3196.97  6368.90 66.57941      199.21676
⋮   ⋮        ⋮        ⋮        ⋮             ⋮        
271  3866.48  1261.04  2605.44 67.38532      206.61042
272  6905.83  4946.42  1959.41 28.37327       39.61269
273  4499.16  2272.93  2226.23 49.48101       97.94538
274 38796.98 23401.67 15395.31 39.68172       65.78723
275 25330.20 10895.38 14434.82 56.98660      132.48570
276  1914.47  1151.36   763.11 39.86012       66.27901
277 40402.43 15868.36 24534.07 60.72424      154.60999
278 24180.51 17008.63  7171.88 29.65975       42.16612
279 32310.35 19328.67 12981.68 40.17809       67.16282
280 11805.27  7370.27  4435.00 37.56797       60.17419
281 22151.01  9576.69 12574.32 56.76635      131.30132
282 18267.85  7013.20 11254.65 61.60906      160.47810
283 15037.78  4788.78 10249.00 68.15501      214.02111
284 49866.51 19117.23 30749.28 61.66319      160.84590
285 20052.34  9685.11 10367.23 51.70085      107.04298
286 17754.91 13390.34  4364.57 24.58233       32.59492
287 31099.55 21502.81  9596.74 30.85813       44.63017
288 15760.95  6493.53  9267.42 58.79988      142.71775
289  3393.75  1052.97  2340.78 68.97326      222.30263
290 11034.63  4603.89  6430.74 58.27780      139.68057
291 27275.19 20480.76  6794.43 24.91066       33.17470
292 23689.77 10952.35 12737.42 53.76760      116.29851
293 26535.57 15430.29 11105.28 41.85054       71.97065
294 34806.29 18741.34 16064.95 46.15531       85.71932
295  2867.68  1632.15  1235.53 43.08465       75.69954
296 38026.53 28410.28  9616.25 25.28827       33.84778
297 10029.25  4614.59  5414.66 53.98868      117.33783
298 37911.51 21905.14 16006.37 42.22034       73.07130
299  2137.30  1156.84   980.46 45.87377       84.75329
300 34730.09 21474.69 13255.40 38.16690       61.72569

## Part 3: Creating Categorical Variables

**Task**: Create categorical variables for business segmentation using conditional logic.

**Business Context**: Segmentation is crucial for targeted business strategies. By categorizing transactions and customers, businesses can tailor their approaches to different performance levels and customer types.

**Categories to create**:
1. **Performance_Category**: Based on Profit_Margin (High: >50%, Medium: 30-50%, Low: <30%)
2. **Revenue_Size**: Based on Revenue (Large: >30000, Medium: 15000-30000, Small: <15000)
3. **Deal_Size**: Based on Units_Sold (Bulk: >50, Standard: 20-50, Small: <20)
4. **High_Value_Customer**: Flag for Revenue > 25000
5. **Profitable_Deal**: Flag for Profit_Margin > 40%

**Hint**: Use `case_when()` for multiple conditions or `ifelse()` for simple binary classifications.

In [ ]:
# Create categorical variables for business segmentation
company_data <- company_data %>%
  mutate(
    # TODO: Create Performance_Category using case_when()
    # High: Profit_Margin > 50, Medium: Profit_Margin > 30, Low: everything else
    Performance_Category = case_when(
      Profit_Margin > 50 ~ "High",
      Profit_Margin > 30 ~ "Medium",
      TRUE ~ "Low"
    ),
    # TODO: Create Revenue_Size category
    # Large: Revenue > 30000, Medium: Revenue > 15000, Small: everything else
    Revenue_Size = case_when(
      Revenue > 30000 ~ "Large",
      Revenue > 15000 ~ "Medium",
      TRUE ~ "Low"
    ),
    # TODO: Create Deal_Size category  
    # Bulk: Units_Sold > 50, Standard: Units_Sold > 20, Small: everything else
    Deal_Size = case_when(
      Units_Sold > 50 ~ "Bulk",
      Units_Sold > 20 ~ "Standard",
      TRUE ~ "Small"
    ),

    # TODO: Create High_Value_Customer flag (Yes/No for Revenue > 25000)
    High_Value_Customer = if_else(Revenue > 25000, "Yes", "No"),
    # TODO: Create Profitable_Deal flag (Yes/No for Profit_Margin > 40)
    Profitable_Deal = if_else(Profit_Margin > 40, "Yes", "No")
  )

# Examine the distribution of categorical variables
# TODO: Use table() to show the distribution of Performance_Category
table(company_data$Performance_Category)

# TODO: Use table() to show the distribution of Revenue_Size
table(company_data$Revenue_Size)

# TODO: Use table() to show the distribution of High_Value_Customer
table(company_data$High_Value_Customer)


  High    Low Medium 
   113     66    121 


 Large    Low Medium 
   122     74    104 


 No Yes 
146 154 

## Part 4: Summary Statistics with summarize()

**Task**: Use `summarize()` to calculate key business metrics and overall performance indicators.

**Business Context**: Executive dashboards and business reports rely on aggregate statistics to provide high-level insights. These summary metrics help stakeholders quickly understand overall business performance.

**Metrics to calculate**:
1. **Total Revenue** and **Total Profit**
2. **Average Profit Margin** and **Average ROI**
3. **Total Units Sold** and **Total Transactions**
4. **Revenue Statistics**: Min, Max, Mean, Median, Standard Deviation
5. **Profit Margin Statistics**: Quartiles and distribution metrics

**Hint**: You can calculate multiple summary statistics in a single `summarize()` call. Use functions like `sum()`, `mean()`, `median()`, `min()`, `max()`, `sd()`, and `n()`.

In [ ]:
# Calculate overall business summary statistics
business_summary <- company_data %>%
  summarize(
    # TODO: Calculate total_revenue (sum of Revenue)
    total_revenue = sum(Revenue),
    # TODO: Calculate total_profit (sum of Profit)
    total_profit = sum(Profit),
    # TODO: Calculate avg_profit_margin (mean of Profit_Margin)
    avg_profit_margin = mean(Profit_Margin),
    # TODO: Calculate avg_roi (mean of ROI)
    avg_roi = mean(ROI),
    # TODO: Calculate total_units (sum of Units_Sold)
    total_units = sum(Units_Sold),
    # TODO: Calculate transaction_count (use n())
    transaction_count = n(),
    # TODO: Calculate avg_revenue_per_transaction (mean of Revenue)
    avg_revenue_per_transaction = mean(Revenue)
  )

# Display the business summary
print("Overall Business Performance Summary:")
print(business_summary)

# Calculate detailed revenue statistics
revenue_statistics <- company_data %>%
  summarize(
    # TODO: Calculate revenue statistics: min, max, mean, median, sd
    min_revenue = min(Revenue),
    max_revenue = max(Revenue),
    mean_revenue = mean(Revenue),
    median_revenue = median(Revenue),
    sd_revenue = sd(Revenue)
  )

print("\nDetailed Revenue Statistics:")
print(revenue_statistics)

[1] "Overall Business Performance Summary:"
# A tibble: 1 × 7
  total_revenue total_profit avg_profit_margin avg_roi total_units
          <dbl>        <dbl>             <dbl>   <dbl>       <dbl>
1      7771711.     3407512.              44.2    93.5       16169
# ℹ 2 more variables: transaction_count <int>,
#   avg_revenue_per_transaction <dbl>
[1] "\nDetailed Revenue Statistics:"
# A tibble: 1 × 5
  min_revenue max_revenue mean_revenue median_revenue sd_revenue
        <dbl>       <dbl>        <dbl>          <dbl>      <dbl>
1       1032.      49956.       25906.         26062.     13944.


## Part 5: Grouped Analysis with group_by()

**Task**: Use `group_by()` combined with `summarize()` to analyze performance across different business dimensions.

**Business Context**: Understanding how different segments perform is crucial for strategic decision-making. Regional analysis helps with resource allocation, product analysis guides inventory decisions, and sales rep analysis informs performance management.

**Analyses to perform**:
1. **Regional Analysis**: Performance by Region
2. **Product Category Analysis**: Performance by Product_Category  
3. **Performance Category Analysis**: Compare High/Medium/Low performers
4. **Sales Rep Analysis**: Individual representative performance

**Metrics for each group**:
- Total Revenue and Profit
- Average Profit Margin
- Transaction Count
- Total Units Sold
- Revenue Share (percentage of total)

**Hint**: Use `group_by()` followed by `summarize()`, and don't forget to use `arrange(desc())` to sort by key metrics.

In [ ]:
# Regional Analysis
regional_performance <- company_data %>%
  group_by(Region) %>%
  summarize(
    # TODO: Calculate total_revenue, total_profit, avg_profit_margin, 
    # transaction_count, total_units for each region
    total_revenue = sum(Revenue),
total_profit = sum(Profit),
avg_profit_margin = mean(Profit_Margin),
transaction_count = n(),
total_units = sum(Units_Sold),
    .groups = 'drop'  # Remove grouping
  ) %>%
  # TODO: Add revenue_share calculation (total_revenue / sum(total_revenue) * 100)
  mutate(
    revenue_share = (total_revenue / sum(total_revenue)) *100
  ) %>%
  # TODO: Arrange by total_revenue in descending order
  arrange(desc(total_revenue))

print("Regional Performance Analysis:")
print(regional_performance)

[1] "Regional Performance Analysis:"


# A tibble: 4 × 7
  Region        total_revenue total_profit avg_profit_margin transaction_count
  <chr>                 <dbl>        <dbl>             <dbl>             <int>
1 Europe             2224182.     1006807.              45.5                82
2 Latin America      2112037.      891481.              43.0                83
3 Asia Pacific       1804243.      827243.              46.0                67
4 North America      1631248.      681981.              42.4                68
# ℹ 2 more variables: total_units <dbl>, revenue_share <dbl>


In [ ]:
# Product Category Analysis
category_performance <- company_data %>%
  group_by() %>%  # TODO: Group by Product_Category
  summarize(
    # TODO: Calculate the same metrics as regional analysis
     total_revenue = sum(Revenue),
    total_profit = sum(Profit),
    avg_profit_margin = mean(Profit_Margin),
    transaction_count = n(),
    total_units = sum(Units_Sold),
    .groups = 'drop'
  ) %>%
  # TODO: Add revenue_share and arrange by total_revenue
  mutate(
    revenue_share = (total_revenue / sum(total_revenue)) * 100
  ) %>%
  arrange(desc(total_revenue))

print("Product Category Performance Analysis:")
print(category_performance)

[1] "Product Category Performance Analysis:"
# A tibble: 1 × 6
  total_revenue total_profit avg_profit_margin transaction_count total_units
          <dbl>        <dbl>             <dbl>             <int>       <dbl>
1      7771711.     3407512.              44.2               300       16169
# ℹ 1 more variable: revenue_share <dbl>


In [ ]:
# Performance Category Analysis (High/Medium/Low performers)
performance_analysis <- company_data %>%
  group_by(Performance_Category) %>%
  summarize(
    # TODO: Calculate summary metrics for each performance level
    total_revenue = sum(Revenue),
    total_profit = sum(Profit),
    avg_profit_margin = mean(Profit_Margin),
    transaction_count = n(),
    total_units = sum(Units_Sold),
    .groups = 'drop'
  ) %>%
  # TODO: Add revenue_share calculation
  mutate(
    revenue_share = (total_revenue / sum(total_revenue)) * 100
  )

print("Performance Category Analysis:")
print(performance_analysis)

[1] "Performance Category Analysis:"
# A tibble: 3 × 7
  Performance_Category total_revenue total_profit avg_profit_margin
  <chr>                        <dbl>        <dbl>             <dbl>
1 High                      2771248.     1682947.              60.1
2 Low                       1813077.      457363.              25.3
3 Medium                    3187386.     1267202.              39.8
# ℹ 3 more variables: transaction_count <int>, total_units <dbl>,
#   revenue_share <dbl>


## Part 6: Advanced Grouping and Cross-Tabulation

**Task**: Perform multi-dimensional analysis and frequency counting using `count()` and advanced grouping.

**Business Context**: Real business questions often require analysis across multiple dimensions simultaneously. For example, "Which product categories perform best in each region?" requires cross-tabulation analysis.

**Analyses to perform**:
1. **Multi-dimensional grouping**: Region × Product Category performance
2. **Count analysis**: Frequency distributions using `count()`
3. **Cross-tabulation**: Performance Category vs Revenue Size
4. **Time-based analysis**: Performance by time periods (if date data available)

**Hint**: Use `count()` for frequency analysis and multiple variables in `group_by()` for cross-dimensional analysis.

In [ ]:
# Multi-dimensional analysis: Region and Product Category
region_category_analysis <- company_data %>%
  group_by() %>%  # TODO: Group by Region and Product_Category
  summarize(
    # TODO: Calculate total_revenue, transaction_count, avg_profit_margin
    total_revenue = sum(Revenue),
    transation_count = n(),
    avg_profit_margin = mean(Profit_Margin),
    .groups = 'drop'
  ) %>%
  # TODO: Arrange by total_revenue descending
  arrange(desc(total_revenue))

print("Top 10 Region-Product Category Combinations:")
print(head(region_category_analysis, 10))

[1] "Top 10 Region-Product Category Combinations:"
# A tibble: 1 × 3
  total_revenue transation_count avg_profit_margin
          <dbl>            <int>             <dbl>
1      7771711.              300              44.2


In [ ]:
# Frequency analysis using count()
print("Frequency Analysis:")

# TODO: Count by Performance_Category
performance_counts <- company_data %>%
  count(Performance_Category)

print("Performance Category Distribution:")
print(performance_counts)

# TODO: Count by Revenue_Size
revenue_size_counts <- company_data %>%
  count(Revenue_Size)

print("Revenue Size Distribution:")
print(revenue_size_counts)

# TODO: Count by Deal_Size
deal_size_counts <- company_data %>%
  count(Deal_Size)

print("Deal Size Distribution:")
print(deal_size_counts)

[1] "Frequency Analysis:"
[1] "Performance Category Distribution:"
# A tibble: 3 × 2
  Performance_Category     n
  <chr>                <int>
1 High                   113
2 Low                     66
3 Medium                 121
[1] "Revenue Size Distribution:"
# A tibble: 3 × 2
  Revenue_Size     n
  <chr>        <int>
1 Large          122
2 Low             74
3 Medium         104
[1] "Deal Size Distribution:"
# A tibble: 3 × 2
  Deal_Size     n
  <chr>     <int>
1 Bulk        159
2 Small        51
3 Standard     90


In [ ]:
# Cross-tabulation analysis
print("Cross-Tabulation Analysis:")

# TODO: Create a cross-tabulation of Performance_Category vs Revenue_Size
# Hint: Use count() with two variables, then consider using pivot_wider() or table()
cross_tab_performance_revenue <- company_data %>%
  count(Performance_Category, Revenue_Size) %>%
  pivot_wider(names_from = Revenue_Size, values_from = n, values_fill = 0)

print("Performance Category vs Revenue Size:")
print(cross_tab_performance_revenue)

# Alternative using base R table function
print("\nUsing table() function:")
# TODO: Create table(company_data$Performance_Category, company_data$Revenue_Size)
table(company_data$Performance_Category, company_data$Revenue_Size)

[1] "Cross-Tabulation Analysis:"
[1] "Performance Category vs Revenue Size:"
# A tibble: 3 × 4
  Performance_Category Large   Low Medium
  <chr>                <int> <int>  <int>
1 High                    41    34     38
2 Low                     28     8     30
3 Medium                  53    32     36
[1] "\nUsing table() function:"


        
         Large Low Medium
  High      41  34     38
  Low       28   8     30
  Medium    53  32     36

## Part 7: Business Intelligence Dashboard

**Task**: Create a comprehensive business intelligence summary that combines all your analysis techniques.

**Business Context**: Executive dashboards consolidate multiple analyses into actionable insights. This section simulates creating a report that would be presented to business stakeholders for strategic decision-making.

**Dashboard sections to create**:
1. **Key Performance Indicators (KPIs)**: Overall business health metrics
2. **Top Performers**: Best regions, categories, and sales representatives
3. **Performance Insights**: Distribution of high/medium/low performers
4. **Efficiency Metrics**: Cost ratios, ROI, and operational efficiency
5. **Strategic Recommendations**: Data-driven business recommendations

**Hint**: Use `cat()` or `print()` statements to create formatted output that resembles a professional business report.

In [ ]:
# Create a comprehensive business intelligence dashboard
cat("\n", "=", rep("=", 60), "\n")
cat("              BUSINESS INTELLIGENCE DASHBOARD\n")
cat("=", rep("=", 60), "\n\n")

# Section 1: Key Performance Indicators
cat("📊 KEY PERFORMANCE INDICATORS\n")
cat("─────────────────────────────────\n")

# TODO: Calculate and display overall KPIs
# - Total Revenue, Total Profit, Overall Profit Margin
# - Total Transactions, Average Deal Size
# - High-Value Customer Rate, Profitable Deal Rate

kpi_summary <- company_data %>%
  summarize(
    # TODO: Add your KPI calculations here
    total_revenue = sum(Revenue, na.rm = TRUE),
    total_profit = sum(Profit, na.rm = TRUE),
    overall_profit_margin = mean(Profit_Margin, na.rm = TRUE),
    total_transactions = n(),
    avg_deal_size = mean(Revenue, na.rm = TRUE),
    high_value_rate = mean(High_Value_Customer == "Yes") * 100,
    profitable_deal_rate = mean(Profitable_Deal == "Yes") * 100
  )

# TODO: Display KPIs using cat() or print() with proper formatting

cat("Total Revenue: $", format(kpi_summary$total_revenue, big.mark = ","), "\n")
cat("Total Profit: $", format(kpi_summary$total_profit, big.mark = ","), "\n")
cat("Overall Profit Margin: ", round(kpi_summary$overall_profit_margin, 2), "%\n")
cat("Total Transactions: ", kpi_summary$total_transactions, "\n")
cat("Average Deal Size: $", format(round(kpi_summary$avg_deal_size, 2), big.mark = ","), "\n")
cat("High-Value Customer Rate: ", round(kpi_summary$high_value_rate, 2), "%\n")
cat("Profitable Deal Rate: ", round(kpi_summary$profitable_deal_rate, 2), "%\n")


# Section 2: Top Performers
cat("\n🏆 TOP PERFORMERS\n")
cat("─────────────────\n")

# TODO: Display top performing region
top_region <- company_data %>%
  group_by(Region) %>%
  summarize(total_revenue = sum(Revenue, na.rm = TRUE), .groups = "drop") %>%
  arrange(desc(total_revenue)) %>%
  slice(1)

  top_category <- company_data %>%
  group_by(Product_Category) %>%
  summarize(total_revenue = sum(Revenue, na.rm = TRUE), .groups = "drop") %>%
  arrange(desc(total_revenue)) %>%
  slice(1)

# TODO: Get the region with highest total revenue
cat("Top Region: ", top_region$Region, " ($", format(top_region$total_revenue, big.mark = ","), ")\n")

# TODO: Display top performing product category  
cat("Top Product Category: ", top_category$Product_Category, " ($", format(top_category$total_revenue, big.mark = ","), ")\n")
# TODO: Get the product category with highest total revenue
perf_dist <- company_data %>%
  count(Performance_Category) %>%
  mutate(percent = round(n / sum(n) * 100, 1))

# TODO: Display performance distribution
cat("\n📈 PERFORMANCE DISTRIBUTION\n")
cat("──────────────────────────\n")
# TODO: Show the count and percentage of High/Medium/Low performers
for (i in 1:nrow(perf_dist)) {
  cat(perf_dist$Performance_Category[i], ": ", perf_dist$n[i], " transactions (", perf_dist$percent[i], "%)\n", sep = "")
}


 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
              BUSINESS INTELLIGENCE DASHBOARD
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

📊 KEY PERFORMANCE INDICATORS
─────────────────────────────────
Total Revenue: $ 7,771,711 
Total Profit: $ 3,407,512 
Overall Profit Margin:  44.22 %
Total Transactions:  300 
Average Deal Size: $ 25,905.7 
High-Value Customer Rate:  51.33 %
Profitable Deal Rate:  56 %

🏆 TOP PERFORMERS
─────────────────
Top Region:  Europe  ($ 2,224,182 )
Top Product Category:  Consulting  ($ 1,978,840 )

📈 PERFORMANCE DISTRIBUTION
──────────────────────────
High: 113 transactions (37.7%)
Low: 66 transactions (22%)
Medium: 121 transactions (40.3%)


## Part 8: Data Validation and Quality Checks

**Task**: Validate your calculated fields and check for data quality issues.

**Business Context**: Data quality is paramount in business analytics. Before presenting results to stakeholders, analysts must verify that calculations are correct and identify any data anomalies that could affect decision-making.

## Part 8: Data Validation and Quality Checks

**Task**: Validate your calculated fields and check for data quality issues.

**Business Context**: Data quality is paramount in business analytics. Before presenting results to stakeholders, analysts must verify that calculations are correct and identify any data anomalies that could affect decision-making. In professional settings, data validation is often the difference between accurate insights and costly business mistakes.

**Why Validation is Critical:**
- **Stakeholder Trust**: Executives need confidence in your analysis
- **Decision Impact**: Business decisions based on incorrect data can be costly
- **Regulatory Compliance**: Many industries require data accuracy documentation
- **Professional Credibility**: Accurate analysis establishes your reputation as a reliable analyst

**Validation Methodology:**
1. **Calculation Verification**: Mathematical accuracy of derived metrics
2. **Range Validation**: Ensure values fall within business-logical ranges
3. **Outlier Detection**: Identify unusual values that might indicate data issues
4. **Consistency Checks**: Verify that related metrics align logically
5. **Missing Value Analysis**: Check for gaps in critical business data
6. **Business Logic Testing**: Ensure results make practical business sense

**Professional Validation Techniques:**
- Use `all.equal()` for precise mathematical verification
- Apply `summary()` to identify outliers and impossible values
- Implement logical tests for business rule compliance
- Cross-check totals and subtotals for internal consistency
- Validate categorical assignments against their criteria

**Hint**: Think like a skeptical executive - question every calculation and assumption. If you can't defend your numbers, neither should your stakeholders trust them.

In [ ]:
# Comprehensive Data Validation and Quality Assurance
cat("📋 COMPREHENSIVE DATA VALIDATION REPORT\n")
cat("════════════════════════════════════════\n\n")

# SECTION 1: Mathematical Calculation Verification
cat("🧮 SECTION 1: CALCULATION VERIFICATION\n")
cat("─────────────────────────────────────────\n")

# 1. Validate profit calculation (Profit = Revenue - Cost)
# TODO: Check if Profit equals Revenue - Cost using all.equal()
profit_validation <- all.equal(company_data$Profit, company_data$Revenue - company_data$Cost)
cat("✓ Profit calculation (Revenue - Cost): ")
cat(ifelse(profit_validation == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# 2. Validate profit margin calculation ((Profit/Revenue)*100)
# TODO: Check if Profit_Margin equals (Profit/Revenue)*100
expected_profit_margin <- (company_data$Profit / company_data$Revenue) * 100
margin_validation <- all.equal(company_data$Profit_Margin, expected_profit_margin)
cat("✓ Profit Margin calculation: ")
cat(ifelse(margin_validation == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# 3. Validate ROI calculation ((Profit/Cost)*100)
# TODO: Check if ROI equals (Profit/Cost)*100
expected_roi <- (company_data$Profit / company_data$Cost) * 100
roi_validation <- all.equal(company_data$ROI, expected_roi)
cat("✓ ROI calculation: ")
cat(ifelse(roi_validation == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# SECTION 2: Range and Outlier Analysis
cat("\n📊 SECTION 2: RANGE AND OUTLIER VALIDATION\n")
cat("──────────────────────────────────────────────\n")

# 4. Check for extreme profit margins (business logic: should rarely exceed 100% or be below -50%)
# TODO: Count transactions with extreme profit margins (>100% or <-50%)
extreme_margins <- sum(company_data$Profit_Margin > 100 | company_data$Profit_Margin < -50, na.rm = TRUE)
cat("⚠️  Extreme profit margins (>100% or <-50%): ", extreme_margins, " transactions\n")

# 5. Check for impossible business values
# TODO: Count negative revenues (impossible in normal business)
negative_revenue <- sum(company_data$Revenue < 0, na.rm = TRUE)
cat("⚠️  Negative revenue values: ", negative_revenue, " transactions\n")

# TODO: Count zero or negative units sold (business logic violation)
invalid_units <- sum(company_data$Units_Sold <= 0, na.rm = TRUE)
cat("⚠️  Invalid units sold (≤0): ", invalid_units, " transactions\n")

# SECTION 3: Missing Data Analysis
cat("\n🔍 SECTION 3: MISSING DATA ANALYSIS\n")
cat("──────────────────────────────────────\n")

# 6. Check for missing values in critical business columns
# TODO: Count missing values in Revenue, Cost, and Units_Sold
cat("Missing value analysis:\n")
missing_revenue <- sum(is.na(company_data$Revenue))
missing_cost <- sum(is.na(company_data$Cost))
missing_units <- sum(is.na(company_data$Units_Sold))
missing_profit <- sum(is.na(company_data$Profit))

cat("  • Missing Revenue values: ", missing_revenue, "\n")
cat("  • Missing Cost values: ", missing_cost, "\n") 
cat("  • Missing Units_Sold values: ", missing_units, "\n")
cat("  • Missing Profit values: ", missing_profit, "\n")

# SECTION 4: Categorical Variable Validation
cat("\n🏷️  SECTION 4: CATEGORICAL VARIABLE VALIDATION\n")
cat("─────────────────────────────────────────────────\n")

# 7. Validate Performance_Category assignments
# TODO: Check if High performers actually have Profit_Margin > 50
high_perf_correct <- sum(company_data$Performance_Category == "High" & company_data$Profit_Margin > 50, na.rm = TRUE)
high_perf_total <- sum(company_data$Performance_Category == "High", na.rm = TRUE)
cat("✓ High Performance category accuracy: ", high_perf_correct, "/", high_perf_total, " correctly assigned\n")

# TODO: Check if Revenue_Size categories match their criteria
large_revenue_correct <- sum(company_data$Revenue_Size == "Large" & company_data$Revenue > 30000, na.rm = TRUE)
large_revenue_total <- sum(company_data$Revenue_Size == "Large", na.rm = TRUE)
cat("✓ Large Revenue_Size category accuracy: ", large_revenue_correct, "/", large_revenue_total, " correctly assigned\n")

# SECTION 5: Summary Statistics Review
cat("\n📈 SECTION 5: SUMMARY STATISTICS REVIEW\n")
cat("─────────────────────────────────────────\n")

cat("Key financial metrics summary (check for reasonableness):\n")
# TODO: Display summary statistics for Profit_Margin and ROI
cat("\nProfit Margin Distribution:\n")
print(summary(company_data$Profit_Margin))

cat("\nROI Distribution:\n")
print(summary(company_data$ROI))

# SECTION 6: Business Logic Validation
cat("\n💼 SECTION 6: BUSINESS LOGIC VALIDATION\n")
cat("─────────────────────────────────────────\n")

# 8. Check if Cost_Ratio + Profit_Margin approximately equals 100%
# TODO: Verify that Cost_Ratio + Profit_Margin ≈ 100% (allowing for rounding)
total_percentage_check <- abs((company_data$Cost_Ratio + company_data$Profit_Margin) - 100)
percentage_errors <- sum(total_percentage_check > 1, na.rm = TRUE)  # Allow 1% tolerance for rounding
cat("✓ Cost_Ratio + Profit_Margin = 100% check: ", percentage_errors, " transactions outside tolerance\n")

# 9. Verify Revenue_Per_Unit and Cost_Per_Unit calculations
# TODO: Check if Revenue_Per_Unit = Revenue / Units_Sold
expected_revenue_per_unit <- company_data$Revenue / company_data$Units_Sold
revenue_per_unit_check <- all.equal(company_data$Revenue_Per_Unit, expected_revenue_per_unit)
cat("✓ Revenue_Per_Unit calculation: ")
cat(ifelse(revenue_per_unit_check == TRUE, "✅ PASSED", "❌ FAILED"), "\n")

# VALIDATION SUMMARY
cat("\n📋 VALIDATION SUMMARY REPORT\n")
cat("═══════════════════════════════\n")
cat("Data validation completed. Review all sections above before proceeding.\n")
cat("\n🎯 Next Steps:\n")
cat("1. Investigate any failed validations or extreme values\n")
cat("2. Document validation results for stakeholder presentation\n")
cat("3. Consider data cleaning if significant issues are found\n")
cat("4. Proceed with analysis only after validation passes\n")
cat("\n⚠️  Remember: Never present unvalidated results to business stakeholders!\n")

📋 COMPREHENSIVE DATA VALIDATION REPORT
════════════════════════════════════════

🧮 SECTION 1: CALCULATION VERIFICATION
─────────────────────────────────────────
✓ Profit calculation (Revenue - Cost): ✅ PASSED 
✓ Profit Margin calculation: ✅ PASSED 
✓ ROI calculation: ✅ PASSED 

📊 SECTION 2: RANGE AND OUTLIER VALIDATION
──────────────────────────────────────────────
⚠️  Extreme profit margins (>100% or <-50%):  0  transactions
⚠️  Negative revenue values:  0  transactions
⚠️  Invalid units sold (≤0):  0  transactions

🔍 SECTION 3: MISSING DATA ANALYSIS
──────────────────────────────────────
Missing value analysis:
  • Missing Revenue values:  0 
  • Missing Cost values:  0 
  • Missing Units_Sold values:  0 
  • Missing Profit values:  0 

🏷️  SECTION 4: CATEGORICAL VARIABLE VALIDATION
─────────────────────────────────────────────────
✓ High Performance category accuracy:  113 / 113  correctly assigned
✓ Large Revenue_Size category accuracy:  122 / 122  correctly assigned

📈 SECTION 5: 

## Part 9: Business Insights and Recommendations

**Task**: Synthesize your analysis into actionable business insights and strategic recommendations.

**Business Context**: The ultimate goal of data analysis is to drive business decisions. This section requires you to think like a business consultant, translating your technical findings into strategic recommendations that stakeholders can act upon.

**Required deliverables**:
1. **Key Findings**: Top 3-5 most important insights from your analysis
2. **Performance Gaps**: Areas where the business is underperforming
3. **Opportunities**: Segments or regions with growth potential  
4. **Strategic Recommendations**: Specific, actionable suggestions
5. **Risk Factors**: Data-driven identification of potential business risks

**Hint**: Frame your insights in business terms, focusing on revenue impact, efficiency improvements, and strategic opportunities rather than just statistical findings.

In [ ]:
# Generate business insights and recommendations
cat("💡 BUSINESS INSIGHTS & RECOMMENDATIONS\n")
cat("════════════════════════════════════════\n\n")

cat("🔍 KEY FINDINGS:\n")
cat("─────────────────\n")

# TODO: Identify and present your top 3-5 key findings

# 1. Which region generates the most revenue? 
top_region_data <- regional_performance %>% slice(1)
cat("1. Top Region: ", top_region_data$Region, " generates $", 
    format(top_region_data$total_revenue, big.mark = ","), 
    " (", round(top_region_data$revenue_share, 1), "% of total revenue)\n")
# 2. What percentage of deals are highly profitable?
("High Value Deal Rate: 51.33%\n")
# 3. Which product category has the best margins?
("Product Category with Highest Margin: Asia Pacific Region - Avg Profit Margin = 46.0\n")
# 4. How many transactions fall into each performance category?
("High Performers: 113 Transactions, Medium Performers: 121 Transactions, Low Performers 66 Transactions = 300 Total Transactions\n")
# Find top performing region


# TODO: Add 3-4 more key findings using your analysis results
("Top Product Category:  Consulting  ($ 1,978,840 )\n")

("Total Profit: $ 3,407,512\n")

("Average Deal Size: $ 25,905.7\n")

cat("\n💰 PERFORMANCE ANALYSIS:\n")
cat("────────────────────────\n")

# TODO: Analyze the distribution of performance categories
# What percentage are high/medium/low performers?
cat("High Performers : 121 transactions (40.3%)
Medium Performers : 113 transactions (37/7%)
Low Performers : 66 Transactions (22.0%)\n")

cat("\n🎯 STRATEGIC RECOMMENDATIONS:\n") 
cat("──────────────────────────────\n")

# TODO: Provide 4-5 specific, actionable recommendations based on your analysis
cat("1. Have less efforts focused in our less profitable market in North America and put more focus on Asia Pacific with it's high margins or Europe with it's highest revenue\n")
cat("2. Have our sales team's goal to raise High Performer transactions to 50% of transactions to raise profits\n")
cat("3. Only Around half of business transactons are profitable so company goals should be to increase that percentage\n")
cat("4. Should raise prices for medium performers to boost profitability\n")
cat("5 Either raise low performers to medium performers or don't waste resources on smaller profits\n")
cat("6. Focus on more bulk orders instead of small ones\n")


cat("\n⚠️  RISK FACTORS & OPPORTUNITIES:\n")
cat("──────────────────────────────────\n")

# TODO: Identify potential risks and opportunities based on your data analysis
cat("1. We have only half of deals bringing in profits to the company:\n")
cat("2. We have a high inconsistency of deal sizes:\n")
cat("3. 66 deals only bringing in $467k in profits are dragging performance down:\n")


💡 BUSINESS INSIGHTS & RECOMMENDATIONS
════════════════════════════════════════

🔍 KEY FINDINGS:
─────────────────
1. Top Region:  Europe  generates $ 2,224,182  ( 28.6 % of total revenue)


[1] "High Value Deal Rate: 51.33%\n"

[1] "Product Category with Highest Margin: Asia Pacific Region - Avg Profit Margin = 46.0\n"

[1] "High Performers: 113 Transactions, Medium Performers: 121 Transactions, Low Performers 66 Transactions = 300 Total Transactions\n"

[1] "Top Product Category:  Consulting  ($ 1,978,840 )\n"

[1] "Total Profit: $ 3,407,512\n"

[1] "Average Deal Size: $ 25,905.7\n"


💰 PERFORMANCE ANALYSIS:
────────────────────────
High Performers : 121 transactions (40.3%)
Medium Performers : 113 transactions (37/7%)
Low Performers : 66 Transactions (22.0%)

🎯 STRATEGIC RECOMMENDATIONS:
──────────────────────────────
1. Have less efforts focused in our less profitable market in North America and put more focus on Asia Pacific with it's high margins or Europe with it's highest revenue
2. Have our sales team's goal to raise High Performer transactions to 50% of transactions to raise profits
3. Only Around half of business transactons are profitable so company goals should be to increase that percentage
4. Should raise prices for medium performers to boost profitability
5 Either raise low performers to medium performers or don't waste resources on smaller profits
6. Focus on more bulk orders instead of small ones

⚠️  RISK FACTORS & OPPORTUNITIES:
──────────────────────────────────
1. We have only half of deals bringing in profits to the company:
2. We have a high i

## Reflection Questions

**Task**: Answer the following questions to demonstrate your understanding of the business applications of data transformation.

**Business Context**: Reflection is a critical part of the learning process in data analytics. Understanding not just how to perform analyses, but when and why to use different techniques, distinguishes proficient analysts from experts.

**Questions to address**:

1. **Technical Mastery**: Which dplyr function (mutate, summarize, group_by, count) did you find most powerful for business analysis, and why?

2. **Business Impact**: What was the most surprising or valuable insight you discovered in your analysis?

3. **Strategic Thinking**: If you were presenting these findings to a company executive, which 3 insights would you prioritize and why?

4. **Methodology**: How did grouping and segmentation help reveal patterns that weren't visible in the raw data?

5. **Future Applications**: What additional business questions could you answer with this dataset using the techniques you've learned?

**Instructions**: Write thoughtful, paragraph-length responses that demonstrate both technical understanding and business acumen.

### Reflection Responses

**1. Most Powerful Function for Business Analysis:**

I found that summarize by far was a powerful dplyr due to the fact it condensed many rows of data to a single value. It was useful when calculating aggregate metrics simultaneously and creating KPI’s for the Dashboard. Along with summarize() when it’s used with group_by() makes it possible to to break the data down ever further. For example when looking at performance across multiple regions this made the process simple when trying to find total profit, total revenue, transaction count and average profit margin.

**2. Most Valuable Business Insight:**

What was the most surprising or valuable insight you discovered in your analysis? An insight that I thought was interesting during my analysis of the dataset was despite having approximately 51% of customers being High-Value only 56% of the total transactions actually generated profit. The reason why I thought this insight was the most surprising was that the High Value customer sales data would lead someone to assume that the business is making high profits from those sales but the 51% of actual profitable sales could tell a different story about profits. In my recommendations I said the company should address the 44% of unprofitable sales such as fixing the profit margins in those sales. 


**3. Top 3 Insights for Executive Presentation:**

  First as addressed in the previous reflection, I would report to the executive that we are having profitability issues with our deals and have 44% of our deals not generating profit, despite having the majority of sales coming from High Value customers. To address this issue we should fix our profit margins with the High Value sales and or change our pricing strategies to lower our unprofitable deals. Next would be that we only have 37.7% of our sales operating at a high performance level meaning that 62.3% deals aren’t being capitalized. To fix this we should focus our resources to raise medium performers to high performers. Lastly would be to focus sales in our highest profit margin region, Asia Pacific, and our highest revenue region, Europe. To accomplish this we should move resources to Asia Pacific and Europe from our underperforming region, North America. The reasoning behind this would be that the North American region has low profit margins, low revenue, and low number of transactions. This shows us that our best ways of maximizing company profits would be to focus on our eastern hemisphere markets. 

**4. Value of Grouping and Segmentation:**

 It allowed for organizing the data into readable categories and values. Connecting multiple functions  in advanced analysis helped for comparison across multiple regions. Showing which regions were under performing, had high profit margins and total revenue. The raw data from the dataset couldn’t show where the company was the strongest and where it needed to improve but using the dplyr functions the data was able to to be simplified to a point where trends could be visible, which I later used in the KPI’s Dashboard. 

**5. Future Business Applications:**

 First “What prices could we charge to generate the most revenue while not overcharging customer?”  We would use a price analysis to do this using count(), mutate(), group_by(), and summarize() we would see how efficiently those products are selling and where the business would need to change prices accordingly. Another question that could be answered with the new dplyr functions I learned would be “Which months generate the most revenue?” using the new dplyr functions we could calculate total monthly revenue, average transaction revenue, and total number of transactions. With this information the prices for the month trends could be changed to maximize our revenue.

## Submission Checklist

Before submitting your homework, ensure you have completed all sections:

### Technical Requirements
- [ ] **Part 1**: Successfully imported data and performed initial exploration
- [ ] **Part 2**: Created all required financial metrics using mutate()
- [ ] **Part 3**: Developed categorical variables for business segmentation  
- [ ] **Part 4**: Generated comprehensive summary statistics
- [ ] **Part 5**: Performed grouped analysis by multiple business dimensions
- [ ] **Part 6**: Conducted frequency analysis and cross-tabulations
- [ ] **Part 7**: Created a professional business intelligence dashboard
- [ ] **Part 8**: Validated data quality and calculation accuracy
- [ ] **Part 9**: Synthesized findings into business insights and recommendations

### Code Quality
- [ ] All code sections are completed with working R code
- [ ] Code is well-commented and follows best practices
- [ ] Variable names are meaningful and consistent
- [ ] Proper use of the pipe operator (%>%) for readable code chains
- [ ] Appropriate use of .groups = 'drop' after group_by operations

### Business Analysis
- [ ] Analysis demonstrates understanding of business context
- [ ] Insights are presented in business-friendly language
- [ ] Recommendations are specific and actionable
- [ ] Cross-dimensional analysis reveals meaningful patterns
- [ ] Dashboard provides comprehensive performance overview

### Reflection and Learning
- [ ] Reflection questions answered thoughtfully and completely
- [ ] Responses demonstrate both technical and business understanding
- [ ] Examples from your own analysis support your reflections

**Final Note**: This homework simulates real-world business analytics work. Focus not just on getting the right technical results, but on developing the analytical thinking skills that drive business value. In professional settings, your ability to translate data into actionable insights is what makes you valuable to an organization.

**Good luck, and remember**: Great analysts don't just manipulate data—they uncover the stories that data tells about business performance and opportunities!

---

## 🚀 Ready to Submit?

### Easy Submission Steps (No Command Line Required!):

1. **Save this notebook** (Ctrl+S or File → Save)

2. **Use VS Code Source Control**:
   - Click the **Source Control** icon in the left sidebar (tree branch symbol)
   - Click the **"+"** button next to your notebook file
   - Type a message: `Submit homework 2 - Data Cleaning - [Your Name]`
   - Click **"Commit"** 
   - Click **"Sync Changes"** or **"Push"**

3. **Verify on GitHub**: Go to your repository online and confirm your notebook appears with your completed work

**📖 Need help?** See [GITHUB_CLASSROOM_SUBMISSION.md](../../GITHUB_CLASSROOM_SUBMISSION.md) for detailed instructions.
